pkg/meta/backup.go

```diff
@@ -0,0 +1,413 @@
+
+	data, err := seg.Marshal()
+	if err != nil {
+		return fmt.Errorf("failed to marshal segment %s: %v", seg, err)
+	}
+
+	n, err := w.Write(data)
+	if err != nil && n != len(data) {
+		return fmt.Errorf("failed to write segment %s: err %v, write len %d, expect len %d", seg, err, n, len(data))
+	}
+
+	name := seg.String()
+	info, ok := f.Footer.Msg.Infos[name]
+	if !ok {
+		info = &pb.Footer_SegInfo{Offset: []uint64{}}
+		f.Footer.Msg.Infos[name] = info
+	}
+
+	info.Offset = append(info.Offset, f.Offset)
+	f.Offset += uint64(n)
+	return nil
+}
+
+func (f *BakFormat) ReadSegment(r io.Reader) (*BakSegment, error) {
+	seg := &BakSegment{}
+	if err := seg.Unmarshal(r); err != nil {
+		return nil, err
+	}
+	return seg, nil
+}
+
+func (f *BakFormat) WriteFooter(w io.Writer) error {
+	if err := f.writeEOS(w); err != nil {
+		return err
+	}
+
+	data, err := f.Footer.Marshal()
+	if err != nil {
+		return err
+	}
+	n, err := w.Write(data)
+	if err != nil && n != len(data) {
+		return fmt.Errorf("failed to write footer: err %v, write len %d, expect len %d", err, n, len(data))
+	}
+	return nil
+}
+
+func (f *BakFormat) writeEOS(w io.Writer) error {
+	if n, err := w.Write(binary.BigEndian.AppendUint32(nil, BakEOS)); err != nil && n != 4 {
+		return fmt.Errorf("failed to write EOS: err %w, write len %d, expect len 4", err, n)
+	}
+	return nil
+}
+
+func (f *BakFormat) ReadFooter(r io.ReadSeeker) (*BakFooter, error) {
+	footer := &BakFooter{}
+	if err := footer.Unmarshal(r); err != nil {
+		return nil, err
+	}
+	if footer.Msg.Magic != BakMagic {
+		return nil, fmt.Errorf("invalid magic number %d, expect %d", footer.Msg.Magic, BakMagic)
+	}
+	return footer, nil
+}
+
+type BakFooter struct {
+	Msg *pb.Footer
+	Len uint64
+}
+
+func (h *BakFooter) Marshal() ([]byte, error) {
+	data, err := proto.Marshal(h.Msg)
+	if err != nil {
+		return nil, fmt.Errorf("failed to marshal footer: %w", err)
+	}
+
+	h.Len = uint64(len(data))
+	data = binary.BigEndian.AppendUint64(data, h.Len)
+	return data, nil
+}
+
+func (h *BakFooter) Unmarshal(r io.ReadSeeker) error {
+	lenSize := int64(unsafe.Sizeof(h.Len))
+	_, _ = r.Seek(lenSize, io.SeekEnd)
+
+	data := make([]byte, lenSize)
+	if n, err := r.Read(data); err != nil && n != int(lenSize) {
+		return fmt.Errorf("failed to read footer length: err %w, read len %d, expect len %d", err, n, lenSize)
+	}
+
+	h.Len = binary.BigEndian.Uint64(data)
+	_, _ = r.Seek(int64(h.Len)+lenSize, io.SeekEnd)
+	data = make([]byte, h.Len)
+	if n, err := r.Read(data); err != nil && n != int(h.Len) {
+		return fmt.Errorf("failed to read footer: err %w, read len %d, expect len %d", err, n, h.Len)
+	}
+
+	if err := proto.Unmarshal(data, h.Msg); err != nil {
+		return fmt.Errorf("failed to unmarshal footer: %w", err)
+	}
+	return nil
+}
+
+type BakSegment struct {
+	Typ uint32
+	Len uint64
+	Val proto.Message
+}
+
+func (s *BakSegment) String() string {
+	return string(proto.MessageName(s.Val).Name())
+}
+
+func (s *BakSegment) Marshal() ([]byte, error) {
```

Comment:
Write the marshaled message into io.Writer directly to avoid memory copy?

Reply:
done